Userland/Libraries/LibJS/Runtime/ECMAScriptFunctionObject.cpp

```diff
@@ -659,13 +667,95 @@ void ECMAScriptFunctionObject::ordinary_call_bind_this(ExecutionContext& callee_
     MUST(verify_cast<FunctionEnvironment>(local_env)->bind_this_value(global_object(), this_value));
 }
 
+// 27.7.5.1 AsyncFunctionStart ( promiseCapability, asyncFunctionBody ), https://tc39.es/ecma262/#sec-async-functions-abstract-operations-async-function-start
+void ECMAScriptFunctionObject::async_function_start(const PromiseCapability& promise_capability)
+{
+    auto& vm = this->vm();
+
+    // 1. Let runningContext be the running execution context.
+    auto& running_context = vm.running_execution_context();
+
+    // 2. Let asyncContext be a copy of runningContext.
+    auto async_context = running_context.copy();
+
+    // 3. NOTE: Copying the execution state is required for AsyncBlockStart to resume its execution. It is ill-defined to resume a currently executing context.
+
+    // 4. Perform ! AsyncBlockStart(promiseCapability, asyncFunctionBody, asyncContext).
+    async_block_start(promise_capability, async_context);
+}
+
+// 27.7.5.2 AsyncBlockStart ( promiseCapability, asyncBody, asyncContext ), https://tc39.es/ecma262/#sec-asyncblockstart
+void ECMAScriptFunctionObject::async_block_start(const PromiseCapability& promise_capability, ExecutionContext& async_context)
+{
+    auto& vm = this->vm();
+
+    // 1. Assert: promiseCapability is a PromiseCapability Record.
+
+    // 2. Let runningContext be the running execution context.
+    auto& running_context = vm.running_execution_context();
+
+    // 3. Set the code evaluation state of asyncContext such that when evaluation is resumed for that execution context the following steps will be performed:
+    auto* execution_steps = NativeFunction::create(global_object(), "", [async_body = m_ecmascript_code, &promise_capability](auto& vm, auto& global_object) -> ThrowCompletionOr<Value> {
+        // NOTE: vm interpreter must exist at this point, since OrdinaryCallEvaluateBody instantiated it
+
+        // a. Let result be the result of evaluating asyncBody.
+        auto result = async_body->execute(*vm.interpreter_if_exists(), global_object);
+
+        // b. Assert: If we return here, the async function either threw an exception or performed an implicit or explicit return; all awaiting is done.
+
+        // c. Remove asyncContext from the execution context stack and restore the execution context that is at the top of the execution context stack as the running execution context.
+        vm.pop_execution_context();
+
+        // NOTE: Running the AST node should eventually return a completion.
+        // Until it does, we assume "return" and include the undefined fallback from the call site.
```

Comment:
This referred to `result.value_or(js_undefined())` where you took the comment from, which doesn't seem to apply here

Reply:
It's not the comment doesn't make sense, I just forgot to add the `.value_or(js_undefined())` :sweat_smile: